### ISSUE TITLE
No default digest for elliptic curve certificates with smime

### ISSUE THREAD

**Issue by hellow554** [openssl/openssl#13523]
I'm using the 1.1.1h release of openssl.
Also I'm using the Version [1.4 version of rauc](https://github.com/rauc/rauc/tree/v1.4) (but it also doesn't work on other version of rauc. Only openssl seems to be important here).

Rauc is a updater for embedded systems. It uses openssl so sign a squashfs file via s/mime. 
When I try to use it with an ed25519 certificate the following error occurs. If I use a RSA certificate, everything goes fine.
The error 

```
Failed to create bundle: failed signing bundle: failed to create signature: error:2E066080:CMS routines:CMS_add1_signer:no default digest
140224897656640:error:2E094063:CMS routines:CMS_sign:add signer error:../openssl-1.1.1h/crypto/cms/cms_smime.c:451:
```

The offending file in rauc is https://github.com/rauc/rauc/blob/ae9cee9f29fc3ba101c887cafe0bca89e4fadd80/src/signature.c#L499-L511

    cms = CMS_sign(signcert, pkey, intercerts, incontent, flags);

**Comment by hellow554** [comment]
```bash
#!/bin/bash

sudo apt install rauc

TEMP=$(mktemp -d)
echo -e '[update]\ncompatible=openssl example\nversion=1.0' > $TEMP/manifest.raucm
yes AU | openssl req -x509 -newkey ED25519 -keyout $TEMP/key.pem -out $TEMP/cert.pem -days 10 -nodes
rauc bundle --debug --cert=$TEMP/cert.pem --key=$TEMP/key.pem $TEMP $(mktemp -u)
```

gives the error from above

**Comment by mattcaswell** [comment]
The "pure" versions of Ed25519/Ed448 (which is what OpenSSL implements) don't use a digest at all, therefore there is no default. It looks like the CMS code doesn't have the necessary support for these certificate types.

**Comment by hellow554** [comment]
Is there any version of ed25519 I could use to work around this bug?

**Comment by mattcaswell** [comment]
No, unfortunately not, the issue is in the CMS code. It needs to be taught how to handle those algorithms.

**Comment by kaduk** [comment]
> Is there any version of ed25519/ed448 I could use to work around this bug?

The short answer is "no", as Matt said already.  Though technically there is a "pre-hashed" version of ed25519/ed448, the so-called ed25519ph/ed448ph, from RFC 8032.  OIDs were allocated for them in RFC 8410, but OpenSSL appears to not actually implement those variants (yet).

**Comment by saper** [comment]
> OIDs were allocated for them in RFC 8410

Did RFC 8410 assign OIDs to pre-hashed variants? [My reading](https://datatracker.ietf.org/doc/draft-ietf-curdle-pkix/shepherdwriteup/) is that it was dropped from the earlier revisions of this document. Also the comment from https://github.com/openssl/openssl/issues/1425

> Rich can correct me if I've got the details wrong, but I believe the current consensus from IETF 96 is:
> * X25519, X448, Ed25519, Ed25519ph, Ed448, and Ed448ph each get distinct top-level AlgorithmIdentifier OIDs, not using id-ecPublicKey, etc.

**Comment by kaduk** [comment]
> > OIDs were allocated for them in RFC 8410
> 
> Did RFC 8410 assign OIDs to pre-hashed variants? [My reading](https://datatracker.ietf.org/doc/draft-ietf-curdle-pkix/shepherdwriteup/) is that it was dropped from the earlier revisions of this document. Also the comment from #1425

That seems correct, good catch and thanks.

**Comment by DDvO** [comment]
See also #11915, where essentially the same issue is being handled.

**Comment by DDvO** [comment]
Feature is in work: draft PR https://github.com/openssl/openssl/pull/22391, help wanted

**Comment by nhorman** [comment]
Given that @DDvO has a draft PR in progress, and is looking for help, I'm marking this as a community issue

**Comment by DDvO** [comment]
Rebased on current `master`, 
This meanwhile includes #26867, which took over part of this PR for a partial solution.
